pkg/controller/sync/version/manager.go

```diff
@@ -0,0 +1,434 @@
+	m.Lock()
+	delete(m.versions, versionQualifiedName.String())
+	m.Unlock()
+	return m.deleteFromAPI(versionQualifiedName)
+}
+
+func (m *VersionManager) list(stopChan <-chan struct{}) (pkgruntime.Object, bool) {
+	targetKind := m.typeConfig.GetTarget().Kind
+
+	// Attempt retrieval of list of versions until success or the channel is closed.
+	var versionList pkgruntime.Object
+	err := wait.PollImmediateInfinite(1*time.Second, func() (bool, error) {
+		select {
+		case <-stopChan:
+			glog.V(4).Infof("Halting version manager list due to closed stop channel")
+			return false, fmt.Errorf("")
+		default:
+		}
+
+		var err error
+		versionList, err = m.adapter.List(m.namespace)
+		if err != nil {
+			runtime.HandleError(fmt.Errorf("Failed to list propagated versions for %q: %v", targetKind, err))
+			// Do not return the error to allow the operation to be retried.
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		return nil, false
+	}
+	return versionList, true
+}
+
+func (m *VersionManager) load(versionList pkgruntime.Object, stopChan <-chan struct{}) bool {
+	targetKind := m.typeConfig.GetTarget().Kind
+	typePrefix := common.PropagatedVersionPrefix(targetKind)
+	ok := m.adapter.Iterate(versionList, func(obj pkgruntime.Object) bool {
+		select {
+		case <-stopChan:
+			glog.V(4).Infof("Halting version manager load due to closed stop channel")
+			return false
+		default:
+		}
+
+		qualifiedName := util.NewQualifiedName(obj)
+		// Ignore propagated version for other types
+		if strings.HasPrefix(qualifiedName.Name, typePrefix) {
+			m.versions[qualifiedName.String()] = obj
+		}
+		return true
+	})
+	if !ok {
+		return false
+	}
+	m.Lock()
+	m.hasSynced = true
+	m.Unlock()
+	glog.V(4).Infof("Version manager for %q synced", targetKind)
+	return true
+}
+
+// versionQualifiedName derives the qualified name of a version
+// resource from the qualified name of a template or target resource.
+func (m *VersionManager) versionQualifiedName(qualifiedName util.QualifiedName) util.QualifiedName {
+	kind := m.typeConfig.GetTarget().Kind
+	namespace := qualifiedName.Namespace
+	if kind == util.NamespaceKind {
+		namespace = qualifiedName.Name
+	}
+	versionName := common.PropagatedVersionName(kind, qualifiedName.Name)
+	return util.QualifiedName{Name: versionName, Namespace: namespace}
+}
+
+// writeVersion serializes the current state of the named propagated version to the API.
+func (m *VersionManager) writeVersion(qualifiedName util.QualifiedName) util.ReconciliationStatus {
+	key := qualifiedName.String()
+	adapterType := m.adapter.TypeName()
+
+	m.RLock()
+	obj, ok := m.versions[key]
+	m.RUnlock()
+	if !ok {
+		// Version is no longer tracked - ensure removal from API
+		err := m.deleteFromAPI(qualifiedName)
+		if err != nil {
+			runtime.HandleError(fmt.Errorf("Failed to delete %s %q: %s", adapterType, key, err))
+			return util.StatusError
+		}
+		return util.StatusAllOK
+	}
+
+	metaAccessor, err := meta.Accessor(obj)
```

Comment:
I'll add a comment why that is not necessary.

Reply:
Thanks, comment is helpful. It might be worth mentioning that the `PropagatedVersion` this object references could still have been removed from the cache in a multi-threaded environment, but we would not get errors thrown because golang will not garbage collect the referenced data until all references are gone, including this one.